### PR TITLE
Add `mixed` parameter and return types to `FilterInterface`

### PR DESCRIPTION
**Pull request by gsteel**
|    Q          |   A
|-------------- | ------
| BC Break      | yes
| QA            | yes

### Description

Adds parameter and return type of `mixed` to `FilterInterface::filter()` and all implementations.

Some additional property/parameter/return types have been added but avoided where possible.

The plan here is to add types everywhere in another patch later on.

### CODE REVIEW

File: psalm-baseline.xml [hunk 1]
```diff
@@ -718,11 +674,14 @@
     <ArgumentTypeCoercion>
       <code><![CDATA[$options['pluginManager']]]></code>
     </ArgumentTypeCoercion>
+    <InvalidArgument>
+      <code>array_values($processedParts)</code>
+    </InvalidArgument>
```
Comment (Ocramius) [inline psalm-baseline.xml]: Anything worth investigating here?

File: src/AbstractDateDropdown.php [hunk 2]
```diff
@@ -18,34 +19,29 @@
...
  */
 abstract class AbstractDateDropdown extends AbstractFilter
 {
     /**
      * If true, the filter will return null if any date field is empty
-     *
-     * @var bool
      */
-    protected $nullOnEmpty = false;
+    protected bool $nullOnEmpty = false;
```
Comment (Ocramius) [inline src/AbstractDateDropdown.php:32]: Not sure if we should add native types to properties: I'm personally OK with it, but it leads to BC breaks downstream, if anybody did override this.

Comment (Ocramius) [reply src/AbstractDateDropdown.php:32]: Ah, never mind, this is for the next major.

At that point, I'd just get rid of `protected`, but that's for another patch.

File: src/AllowList.php [hunk 3]
```diff
@@ -85,7 +85,7 @@ public function getList()
      *
      * Will return $value if its present in the white-list. If $value is rejected then it will return null.
      */
-    public function filter($value)
+    public function filter(mixed $value): mixed
```
Comment (Ocramius) [inline src/AllowList.php:93]: Could be refined to return one of the values of the specified list, or `null`

File: src/Boolean.php [hunk 4]
```diff
@@ -206,9 +206,8 @@ public function getTranslations()
      * Returns a boolean representation of $value
      *
      * @param  null|array|bool|float|int|string $value
```
Comment (Ocramius) [inline src/Boolean.php:209]: Lies :D

File: src/Decompress.php [hunk 5]
```diff
@@ -29,10 +29,10 @@ public function __invoke($value)
...
      *
-     * @param  string $value Content to decompress
-     * @return string The decompressed content
+     * @param mixed $value Content to decompress
+     * @return mixed The decompressed content
      */
-    public function filter($value)
+    public function filter(mixed $value): mixed
     {
         if (! is_string($value) && $value !== null) {
```
Comment (Ocramius) [inline src/Decompress.php:39]: wait, we decompress `null`? :thinking:

**Review by Ocramius — CHANGES_REQUESTED** [review]
Some changes look like scope creep in here :D

File: psalm-baseline.xml (hunk 1 above)
Comment (gsteel) [reply psalm-baseline.xml]: Nah. This is psalm better understanding how little it understands this existing code. My current plan is to get native types everywhere and whittle down the baseline as much as poss for v3, so it's something I should get to later on.

File: src/Decompress.php (hunk 5 above)
Comment (gsteel) [reply src/Decompress.php:39]: I'll have to look closer at this later on, but I think that some of the adapters only work on files passed as options (Those adapters might be dead now), therefore there's no argument to `filter()`, or something like that…

**Comment by gsteel** [comment]
@Ocramius

> Some changes look like scope creep in here :D

Not sure I understand this… This patch is just meant to nail down the contract of `filter()` to `filter(mixed): mixed` with minimal other changes. I'm planning on going through the filters after and adding BC breaking types everywhere, along with `final` and improving type inference where possible. Just trying my hardest to keep the patch size as small as possible 😬

**Review by Ocramius — APPROVED** [review]
@gsteel my bad: I think I initially reviewed this patch with the mindset that it was for a stable branch, but I think all is good here.

**Pull request merged**